### PR TITLE
User empty cart #45

### LINKED ISSUE #44: User Story 29, User can manipulate quantities in their cart
```
As a visitor or registered user
When I have items in my cart
And I visit my cart
Next to each item in my cart
I see a button or link to remove that item from my cart
- clicking this button will remove the item but not other items

I see a button or link to increment the count of items I want to purchase
- I cannot increment the count beyond the merchant's inventory size

I see a button or link to decrement the count of items I want to purchase
- If I decrement the count to 0 the item is immediately removed from my cart
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #49

child of #79

depends on #47

### LINKED ISSUE #45: User Story 28, User can empty a cart that has items
```
As a visitor or registered user
When I have items in my cart
And I visit my cart ("/cart")
And I click the link to empty my cart
Then I am returned to my cart
All items have been completely removed from my cart
The navigation bar shows 0 items in my cart
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #49

child of #79

depends on #47

### PR DESCRIPTION
**Pull request by allieIsOkay**
- Adds tests for incrementing, decrementing, and removing items from cart show page
- Adds tests for Cart instance methods to decrement and remove all of an item
- Adds functionality for cart class to decrement or remove items
- Adds display of functionality to cart show page
- Updates navbar to be sticky-top
- Adjusts organization of cart show page to prevent duplication of dividers when looping

- closes #45 
- closes #44

### CODE REVIEW

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -27,8 +27,35 @@ def create
...
+
+  def destroy
+    @cart.contents.clear
+    redirect_to cart_path
+  end
+
   private
 
+  def adjust_quantity
+    if @new_quantity > 0
```
Comment (allieIsOkay) [inline app/controllers/carts_controller.rb:48]: Couldn't determine a better scheme for handling the multitude of operations, so diverted logic to a helper method here to handle the conditionals for updating the cart. If anyone has a better Idea i'm open to it.

File: app/controllers/carts_controller.rb [hunk 2]
```diff
@@ -27,8 +27,35 @@ def create
...
+      if @new_quantity < @current_quantity
+        @cart.remove_item(@item_id_str)
+      elsif @new_quantity <= @item.quantity
```
Comment (allieIsOkay) [inline app/controllers/carts_controller.rb:51]: Was unable to reuse create function due to existing redirect to item index page.

File: app/views/carts/show.html.erb [hunk 3]
```diff
@@ -1,39 +1,36 @@
...
         <h4 class="product-name"><strong> <%= item.name %> </strong></h4>
         <h4><small> <%= item.description %> </small></h4>
         <h4><small> <strong>Merchant: </strong><%= item.user.name %> </small></h4>
       </div>
-      <div class="col-md-4">
-          <p><strong>Price:</strong>  <%= number_to_currency(item.price)%></p>
-          <p><strong>Quantity:</strong> <%= quantity %></p>
-          <p><strong>Subtotal:</strong> <%= number_to_currency(item.subtotal(quantity)) %></p>
-
+      <div class="col-md-3">
```
Comment (allieIsOkay) [inline app/views/carts/show.html.erb:12]: Adjusted grid width here slightly to allow buttons to display on right side if larger display is being used.

screen shot 2019-02-17 at 8 29 14 pm

File: app/views/shared/_nav.html.erb [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<nav class="navbar navbar-expand-lg navbar-light bg-light">
+<nav class="navbar sticky-top navbar-expand-lg navbar-light bg-light">
```
Comment (allieIsOkay) [inline app/views/shared/_nav.html.erb:1]: Added sticky-top to nav to keep nav present on page when scrolling.

File: config/routes.rb [hunk 5]
```diff
@@ -8,6 +8,8 @@
   resources :users, only: [:create, :update]
 
   get '/cart', to: 'carts#show'
+  delete '/cart', to: 'carts#destroy'
+  put '/cart', to: 'carts#update'
```
Comment (allieIsOkay) [inline config/routes.rb:12]: Used custom routes to circumvent :id requirement on routing.

**Review by allieIsOkay — COMMENTED** [review]
Left some comments regarding changes present, I'd love some feedback on these items.

**Review by timnallen — APPROVED** [review]
Great work Peter. I have some cart issues from what I was working on that may require us to change some things, but I'll go over that with you guys tomorrow and maybe there's a better way.

File: app/views/carts/show.html.erb (hunk 3 above)
Comment (teresa-m-knowles) [reply app/views/carts/show.html.erb:12]: That makes sense, thanks!

File: app/controllers/carts_controller.rb (hunk 1 above)
Comment (csvlewis) [reply app/controllers/carts_controller.rb:48]: I think it's fine, we went over this together

**Review by csvlewis — APPROVED** [review]
